File: src/shared/models/gpu.py
"""GPU domain models.

Spec Reference: specs/01-data-models.md Section 4 - GPU Domain Models
"""

from datetime import datetime
from enum import Enum
from uuid import UUID

from pydantic import Field

from .base import AIOpsBaseModel


class GPUProcessType(str, Enum):
    """Type of GPU process. Spec: Section 4.3"""

    COMPUTE = "COMPUTE"
    GRAPHICS = "GRAPHICS"
    MIXED = "MIXED"


class GPUProcess(AIOpsBaseModel):
    """A process running on a GPU.

    Spec Reference: Section 4.3
    """

    pid: int
    process_name: str
    used_memory_mb: int
    type: GPUProcessType


class GPU(AIOpsBaseModel):
    """A single GPU device.

    Spec Reference: Section 4.2
    """

    index: int = Field(description="GPU index on the node")
    uuid: str = Field(description="NVIDIA GPU UUID")
    name: str
    driver_version: str
    cuda_version: str | None = None
    memory_total_mb: int
    memory_used_mb: int
    memory_free_mb: int
    utilization_gpu_percent: int = Field(ge=0, le=100)
    utilization_memory_percent: int = Field(ge=0, le=100)
    temperature_celsius: int
    power_draw_watts: float
    power_limit_watts: float
    fan_speed_percent: int | None = Field(default=None, ge=0, le=100)
    processes: list[GPUProcess] = Field(default_factory=list)


class GPUNode(AIOpsBaseModel):
    """A node with GPU(s) in a cluster.

    Spec Reference: Section 4.1
    """

    cluster_id: UUID
    cluster_name: str
    node_name: str
    gpus: list[GPU] = Field(default_factory=list)
    last_updated: datetime

File: src/frontend/src/types/gpu.ts
export interface GPUMetrics {
  utilization_percent: number;
  memory_used_bytes: number;
  memory_total_bytes: number;
  temperature_celsius: number;
  power_usage_watts: number;
  power_limit_watts: number;
  clock_speed_mhz: number;
  memory_clock_mhz: number;
}

export interface GPUProcess {
  pid: number;
  name: string;
  memory_used_bytes: number;
  gpu_utilization_percent: number;
}

export interface GPU {
  index: number;
  uuid: string;
  name: string;
  serial: string | null;
  pci_bus_id: string;
  metrics: GPUMetrics;
  processes: GPUProcess[];
}

export interface GPUNode {
  cluster_id: string;
  cluster_name: string;
  node_name: string;
  gpus: GPU[];
  driver_version: string;
  cuda_version: string;
  collected_at: string;
}

export interface GPUSummary {
  total_nodes: number;
  total_gpus: number;
  total_memory_gb: number;
  used_memory_gb: number;
  avg_utilization_percent: number;
  gpu_types: Record<string, number>;
  clusters_with_gpu: number;
}
